### PR TITLE
MINOR: Don't include checker-qual in distribution

### PR DESCRIPTION
**Pull request by mimaison**
checker-qual is only used at compile time, we should not ship it as part of our distribution

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by showuon — COMMENTED** [review]
Traced back to Caffeine project, I saw this issue: https://github.com/ben-manes/caffeine/issues/300#issuecomment-507292996 . In summary:

1. Not just `checker-qual`, we also need to exclude `error_prone_annotations` lib.
2. From the author of Caffeine project:
> checker-qual And errorprone are annotations which Java doesn’t require at runtime unless you reflectively load them when inspecting the methods. Unfortunately other languages will not classload without them, making them required. You can exclude safely in Java.

From my understanding, as long as we can build Kafka successfully with scala 2.12/2.13 after excluding these 2 libs, it should be fine.

**Comment by showuon** [comment]
Here's another explanation from the author of Caffeine: https://github.com/ben-manes/caffeine/issues/474#issuecomment-728243942 to have a replacement for jsr-305. I think it's safe to exclude them for us since the issue should be fixed in scala now. (hope so!)

**Comment by mimaison** [comment]
I tried removing `error_prone_annotations` using:
```
implementation(libs.caffeine) {
      exclude group: 'org.checkerframework', module: 'checker-qual'
      exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
}
```

 but then compilation is failing:
```
> Task :storage:compileJava
/Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Cache.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith': class file for com.google.errorprone.annotations.CompatibleWith not found
/Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Cache.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith'
error: warnings found and -Werror specified
/Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Policy.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith'
```

I've got to admit I've not looked into it further and decided to just remove `checker-qual`.

**Comment by showuon** [comment]
> I tried removing `error_prone_annotations` using:
> 
> ```
> implementation(libs.caffeine) {
>       exclude group: 'org.checkerframework', module: 'checker-qual'
>       exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
> }
> ```
> 
> but then compilation is failing:
> 
> ```
> > Task :storage:compileJava
> /Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Cache.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith': class file for com.google.errorprone.annotations.CompatibleWith not found
> /Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Cache.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith'
> error: warnings found and -Werror specified
> /Users/mickael/.gradle/caches/modules-2/files-2.1/com.github.ben-manes.caffeine/caffeine/2.9.3/b162491f768824d21487551873f9b3b374a7fe19/caffeine-2.9.3.jar(/com/github/benmanes/caffeine/cache/Policy.class): warning: Cannot find annotation method 'value()' in type 'CompatibleWith'
> ```
> 
> I've got to admit I've not looked into it further and decided to just remove `checker-qual`.

Hmm... I guess it is because we're using 2.x of Caffeine, not 3.x. OK, let's exclude `checker-qual` only.

**Review by showuon — APPROVED** [review]
LGTM!

File: build.gradle [hunk 1]
```diff
@@ -1931,7 +1931,9 @@ project(':storage') {
     implementation project(':storage:storage-api')
     implementation project(':server-common')
     implementation project(':clients')
-    implementation libs.caffeine
+    implementation(libs.caffeine) {
+      exclude group: 'org.checkerframework', module: 'checker-qual'
```
Comment (chia7712) [inline build.gradle:1935]: It seems `error_prone_annotations` is used in compile time, right? If so, should it be excluded too?

**Review by chia7712 — COMMENTED** [review]
@mimaison thanks for this patch. Should we remove license file also?

https://github.com/apache/kafka/blob/trunk/licenses/checker-qual-MIT

File: build.gradle (hunk 1 above)
Comment (showuon) [reply build.gradle:1935]: @chia7712 , we've discussed `error_prone_annotations` above. In short, it'll fail compilation if exclude it.

**Comment by ben-manes** [comment]
> warning: Cannot find annotation method 'value()' in type 'CompatibleWith'

I think you could use `-Xlint:-processing` or `-Xlint:-classfile` to disable that warning. There isn't a targeted way to suppress it that I am aware of, and experience it as well due to dependencies excluding annotations that javac wants. For Caffeine's build, when JMH creates its fat jar then Apache's Jackrabbit relocation of Guava emits this warning due to them not relocating the annotations along with it. It's about potentially missing classes for compile-time annotation processors, which if fair if they would query for it but is not useful otherwise. It would be nice if javac could allow more targeted suppressions to avoid disabling the entire category.

> Hmm... I guess it is because we're using 2.x of Caffeine, not 3.x. 

When reasonable, please use the latest release.

**Comment by showuon** [comment]
> When reasonable, please use the latest release.

@ben-manes , thanks for the suggestion. But since we're still supporting jdk8, we can't upgrade to 3.x version.

**Review by satishd — APPROVED** [review]
Thanks @mimaison for the PR, overall LGTM. As @chia7712 suggested earlier, can we also remove the file `checker-qual-MIT` from `licenses` dir?

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by mimaison** [comment]
Test failures are not related, merging to trunk

**Pull request merged**